fabio/GEimage.py

```diff
@@ -1,332 +1,384 @@
+    ('SetConvertLowTime', 2, '<H'),
+    ('SetSyncHighTime', 2, '<H'),
+    ('SetEOLTime', 2, '<H'),
+    ('SetRampOffsetTime', 2, '<H'),
+    ('FOVStartingValue', 2, '<H'),
+    ('ColumnBinning', 2, '<H'),
+    ('RowBinning', 2, '<H'),
+    ('BorderColumns64', 2, '<H'),
+    ('BorderRows64', 2, '<H'),
+    ('FETOffRows64', 2, '<H'),
+    ('FOVStartColumn128', 2, '<H'),
+    ('FOVStartRow128', 2, '<H'),
+    ('NumberOfColumns128', 2, '<H'),
+    ('NumberOfRows128', 2, '<H'),
+    ('VFPAquisition', 2000, None),
+    ('Comment', 200, None)]
+
+
+class GeImage(FabioImage):
+    """GE a-Si Angio detector file format."""
+
+    DESCRIPTION = "GE a-Si Angio detector file format"
+
+    DEFAULT_EXTENSIONS = []
+
+    _need_a_seek_to_read = True
+
+    # need these defaults for the case of the bastardized APS files
+    _HeaderNBytes = 8192
+    _UserHeaderSizeInBytes = 0
+    _NumberOfRowsInFrame = 2048
+    _NumberOfColsInFrame = 2048
+    _ImageDepthInBytes = 2
+    _ImageDepthInBits = 8 * _ImageDepthInBytes
+    _BytesPerFrame = _ImageDepthInBytes \
+        * _NumberOfRowsInFrame \
+        * _NumberOfRowsInFrame
+
+    def _readheader(self, infile):
+        """Read a GE image header."""
+        # !!!: At APS they blanked the header with 8192 bytes of zeros when
+        #      updating the GE firmware.  This happened in ~2018 to the best
+        #      of my knowliedge, and *a lot* of data has been measured with
+        #      the GE detectors that are missing the magic bytes and header.
+        #      This hacky fix works, and we might be stuck with it.
+        aps_img_format = b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+
+        infile.seek(0)
+
+        self.header = self.check_header()
+        for name, nbytes, fmt in GE_HEADER_INFO:
+            if fmt is None:
+                self.header[name] = infile.read(nbytes)
+            else:
+                self.header[name] = struct.unpack(fmt, infile.read(nbytes))[0]
+        if self.header["ImageFormat"] == aps_img_format:
+            # FIXME: for now have to assume we have a bastardized GE file from
+            # Argonne, where they blanked the header with all zeros
+            self.header['StandardHeaderSizeInBytes'] = self._HeaderNBytes
+            self.header['UserHeaderSizeInBytes'] = self._UserHeaderSizeInBytes
+            self.header['NumberOfRowsInFrame'] = self._NumberOfRowsInFrame
+            self.header['NumberOfColsInFrame'] = self._NumberOfColsInFrame
+            self.header['ImageDepthInBits'] = self._ImageDepthInBits
+
+    def read(self, fname, frame=None):
+        """Read header into self.header and the data into self.data."""
+        if frame is None:
+            frame = 0
+        self.header = self.check_header()
+        self.resetvals()
+        infile = self._open(fname, "rb")
+        self.sequencefilename = fname
+        self._readheader(infile)
+
+        # obsolete for now # self.nframes = self.header['NumberOfFrames']
```

Comment:
What if header was not corrupted? How about:
```
if self.header['NumberOfFrames'] == 0 : # struct.unpack("<H", b"\x00\x00")[0]
    file_size = os.stat(fname).st_size
    assert (numpy.remainder(file_size, self._BytesPerFrame)
                == self._HeaderNBytes), "GE file size is incorrect"
    self._nframes = file_size // self._BytesPerFrame
else:
    self._nframes = self.header['NumberOfFrames']

Reply:
That is a more elegant solution for sure -- I agree.